Enable NullAway in multiple projects

reinsch82 opened the pull request:
### Context

Part of the ongoing effort to enable NullAway across the codebase project by project (in dependency-topological order), with one commit per project.

This PR enables NullAway (`errorprone { nullawayEnabled = true }`) in 11 projects:

- `:internal-instrumentation-processor`
- `:service-registry-builder`
- `:java-api-extractor`
- `:snapshots`
- `:wrapper-shared`
- `:wrapper-main`
- `:build-cache-packaging`
- `:normalization-java`
- `:file-watching`
- `:native`
- `:build-cache`

### Approach

Annotations-only — no behavior changes (one deliberate exception below):

- `@Nullable` on genuinely optional fields, parameters, and returns (matching already-annotated interfaces where overrides were missing it).
- `requireNonNull(...)` where a value is non-null in correct usage but NullAway cannot prove it (e.g. set-after-construction fields, map lookups known to be present).
- Targeted `@SuppressWarnings("NullAway")` with a justifying comment only where neither fits.

### Notable points for review

- **Wrapper nullability** (`:wrapper-shared`): `WrapperConfiguration.getDistribution()` stays `@Nullable` — a null distribution is a legitimate, documented state (`WrapperExecutor.forProjectDirectory()` without a wrapper properties file; `DistributionFactory` relies on it to fall back to the current Gradle version). `requireNonNull` moved into the callers that genuinely require a configured distribution (`Install.createDist()`, `PathAssembler.getDistribution()`). The setter accepts null to satisfy the accessor-symmetry rule in `KotlinCompatibilityTest`.
- **Wrapper JAR hash**: `EXPECTED_WRAPPER_JAR_HASH` in `WrapperGenerationIntegrationTest` is updated — the `@Nullable` annotations added to wrapper classes intentionally change the `gradle-wrapper.jar` bytecode.
- **NullAway #681 cleanup**: NullAway now infers that `AtomicReference` holds a non-null type (uber/NullAway#681), so the pre-existing workarounds (`requireNonNull` calls and suppressions) are removed in `DefaultBuildCacheController`, `BaseLocalBuildCacheServiceHandle`, `FileWatchingFilter`, `DefaultBuildOperationListenerManager`, `DefaultWorkerLeaseService`, `DefaultServiceRegistry` (×2), and `InstrumentedInputs`.
- **New `@Initializer` annotation** (`:java-api-extractor`): `ApiMemberSelector.classMember` is populated by the ASM `visit()` callback; NullAway treats any annotation named `Initializer` as marking a field initializer, which replaces a blanket suppression.
- **`FieldMember.compareTo` suppression restored**: the Linux CI toolchain no longer flags it, but the Windows ARM toolchain still rejects passing the `@Nullable` `value` fields to `Comparator.compare()` without an explicit `@Nullable` type argument — the previously reviewed suppression is kept.

### Reviewing cheatsheet

Before merging the PR, comments starting with
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

reinsch82 commented:
@mlopatkin rebased this one too.

mlopatkin commented on platforms/core-configuration/java-api-extractor/src/main/java/org/gradle/internal/tools/api/impl/ApiMemberSelector.java:
```diff
@@ -54,6 +54,7 @@ public class ApiMemberSelector extends ClassVisitor {
     private final boolean apiIncludesPackagePrivateMembers;
 
     private boolean isInnerClass;
+    @SuppressWarnings("NullAway")
```
🤔 This can probably be addressed by an [`@Initializer`-style annotation](https://github.com/uber/NullAway/wiki/Supported-Annotations#initialization).

Could you please add a TODO here?

mlopatkin commented on platforms/core-configuration/java-api-extractor/src/main/java/org/gradle/internal/tools/api/impl/SortingAnnotationVisitor.java:
```diff
@@ -28,6 +28,7 @@ public class SortingAnnotationVisitor extends AnnotationVisitor {
     private final List<AnnotationValue<?>> annotationValues = new LinkedList<>();
     private final AnnotationMember annotation;
 
+    @SuppressWarnings("NullAway")
```
❓ Why is it suppressed rather than made `@Nullable`?

mlopatkin commented on platforms/core-execution/build-cache/src/main/java/org/gradle/caching/internal/controller/service/BaseLocalBuildCacheServiceHandle.java:
```diff
@@ -46,7 +47,7 @@ public LocalBuildCacheService getService() {
         AtomicReference<Optional<BuildCacheLoadResult>> result = new AtomicReference<>(Optional.empty());
         service.loadLocally(key, file -> result.set(Optional.ofNullable(unpackFunction.apply(file))));
-        return result.get();
+        return Cast.unsafeStripNullable(result.get());
```
💅 Could you please add a TODO referencing https://github.com/uber/NullAway/issues/681? I believe it should be fixed in the newer versions of nullaway.

mlopatkin commented on platforms/core-execution/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/NonHierarchicalFileWatcherUpdater.java:
```diff
@@ -159,7 +159,7 @@ private void updateWatchedDirectories(Map<String, Integer> changedWatchDirectori
             }
         } catch (NativeException e) {
-            if (e.getMessage().contains("Already watching path: ")) {
+            if (e.getMessage() != null && e.getMessage().contains("Already watching path: ")) {
```
🤔 Is NullAway smart/dumb enough to believe that `getMessage` always returns the same result?

mlopatkin commented on platforms/core-execution/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/FileWatchingFilter.java:
```diff
@@ -57,6 +57,14 @@ public void locationsWritten(Iterable<String> locations) {
+     * FileHierarchySet.empty() and always set to non-null values in locationsWritten()
+     * and resetLocationsWritten().</p>
+     */
+    @SuppressWarnings("NullAway")
```
💅 Could you please add a TODO referencing https://github.com/uber/NullAway/issues/681?

mlopatkin commented on platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/modelreader/impl/TypeUtils.java:
```diff
@@ -118,6 +118,7 @@ public static List<ExecutableElement> getExecutableElementsFromElements(Stream<?
     public static String elementQualifiedName(Element element) {
         if (element instanceof ExecutableElement) {
+            @SuppressWarnings("NullAway")
             String enclosingTypeName = ((TypeElement) element.getEnclosingElement()).getQualifiedName().toString();
```
❓ Can we replace suppression with `requireNonNull`?

mlopatkin reviewed (changes requested):
Great work! I left a few comments. It also looks like some tests are now legitimately failing.

I think the PR can target master now, and we can merge it as soon as it's green.

reinsch82 commented:
@mlopatkin reopened and addressed your review comments in 7989d9bb668:

- **ApiMemberSelector** (`classMember`): added a TODO to replace the suppression with an `@Initializer`-style annotation.
- **SortingAnnotationVisitor** (`parentVisitor`): made it `@Nullable` and `requireNonNull` it at the single dereference, instead of suppressing the field.
- **FieldMember.compareTo**: kept the suppression but documented why both it and the explicit `@Nullable` type argument are needed — the type argument is required to compare the nullable `value` fields, but it makes NullAway reject `Ordering.arbitrary()` (a `Comparator<Object>`) for a `@Nullable` type variable.
- **TarBuildCacheEntryPacker**: replaced Guava's `checkNotNull` with `Objects.requireNonNull`.
- **OpFiringRemoteBuildCacheServiceHandle.getIOException**: the cause is never null (the only constructor takes a non-null cause), so it now `requireNonNull`s instead of returning `@Nullable`.
- **DefaultBuildCacheController** & **BaseLocalBuildCacheServiceHandle**: added TODOs referencing https://github.com/uber/NullAway/issues/681.
- **FileWatchingFilter**: added a TODO referencing NullAway#681.
- **TypeUtils.elementQualifiedName**: replaced the suppression with `requireNonNull`.
- **NonHierarchicalFileWatcherUpdater**: already resolved in our earlier discussion (plain `getMessage() != null` check, no suppression).

reinsch82 commented:
@mlopatkin I need to rebase this...

reinsch82 committed:
Address some nullability issues in :internal-instrumentation-processor

Enable NullAway and resolve findings, mostly:
- requireNonNull on Element.getEnclosingElement() (always present for the
  method/field elements processed here) and on Map lookups known to be present;
- @Nullable on genuinely-optional fields/params/returns (projectName,
  BridgedMethodInfo, SignatureTree leaf/children, classNameForRequest,
  callableKindForJavadoc, GradleLazyType.className).

reinsch82 committed:
Enable NullAway in :service-registry-builder

Mark the optional builder fields displayName and scope @Nullable, and add a
matching @Nullable to ScopedServiceRegistry's displayName parameter (it forwards
to the already-@Nullable DefaultServiceRegistry constructor).

reinsch82 committed:
Enable NullAway in :java-api-extractor

- Move @Nullable to the array position (String @Nullable []) for the ASM visitor
  interfaces/exceptions parameters; mark MethodMember.signature and
  JavaApiMemberWriter.writeModule's version @Nullable.
- SortingAnnotationVisitor.parentVisitor: @Nullable with requireNonNull at the
  single dereference.
- ApiMemberSelector.classMember: suppress with a TODO to use an @Initializer-style
  annotation (set by the ASM visit() callback before use).

reinsch82 committed:
Enable NullAway in :snapshots

Mark genuinely-optional fields/returns/params @Nullable (lazy hash, builder
result/leaveDirectory/getResult, RelativePathTracker.rootName, predicate,
UnknownImplementationSnapshot.getClassLoaderHash, equals(@Nullable Object)).
Use requireNonNull where a value is non-null in correct usage but NullAway
cannot prove it (RelativePathTracker.leave, PathVisitor.getResult).

reinsch82 committed:
Enable NullAway in :wrapper-shared

- Mark genuinely-optional params/fields @Nullable: Download progressListener and
  DefaultDownloadProgressListener.delegate (null-checked), WrapperConfiguration's
  optional distributionSha256Sum, Install's verifyDownloadChecksum.expectedSum and
  InstallCheck.gradleHome/failureMessage, ExclusiveFileAccessManager.maybeCloseQuietly,
  WrapperExecutor.defaultValue, BootstrapMainStarter.findLauncherJar (caller null-checks),
  Install.fetchDistributionSha256Sum.
- requireNonNull where a value is required but set after construction / non-null in
  context: WrapperConfiguration.getDistribution and InstallCheck.gradleHome after isVerified().
- Suppress WrapperExecutor.getProperty (returns null only in the !required branch).

reinsch82 committed:
Enable NullAway in :wrapper-main

reinsch82 committed:
Enable NullAway in :build-cache-packaging

Use requireNonNull on the per-tree snapshot lookup in pack() (every visited output
tree has a corresponding snapshot in the map).

reinsch82 committed:
Enable NullAway in :normalization-java

- Annotate the ResourceHasher hash() overrides @Nullable (interfaces are already
  @Nullable: null means ignore the resource); mark the FallbackStrategy IoFunction
  type argument @Nullable HashCode and CachingZipEntry.content byte @Nullable [].
- requireNonNull on the manifest keyset-then-get lookup; suppress NullAway on
  ZipHasher's result-discarding withInputStream callback.

reinsch82 committed:
Enable NullAway in :file-watching

- Mark stateful/optional fields @Nullable (WatchingVirtualFileSystem.watchRegistry,
  reasonForNotWatchingFiles, MutableFileWatchingStatistics.errorWhileReceivingFileChanges)
  and the getStatistics() overrides @Nullable to match their interfaces.
- NonHierarchicalFileWatcherUpdater: null-check getMessage() instead of suppressing.
- requireNonNull on watchRegistry derefs inside lambdas where the null-check narrowing
  is lost; suppress FileWatchingFilter.shouldWatchLocation with a TODO referencing
  NullAway#681 (AtomicReference.get()).

reinsch82 committed:
Enable NullAway in :native

- @Nullable on console-detector getConsole() overrides (interface already @Nullable),
  JansiStorageLocator.locate(), EnvironmentModificationResult.reason, GenericFileSystem.caseSensitive,
  ProcessEnvironment.maybeGetPid()/its impl + pid field.
- null-check getMessage() in Jdk7Symlink; requireNonNull where set-before-use or required
  (GenericFileSystem isCaseSensitive/hasContent).
- Suppress EnvironmentModificationResult.toString() (nullable reason) and the NativeServices
  constructor (guarded native-exception inspection); NullAway.Init on the static instance.

reinsch82 committed:
Enable NullAway in :build-cache

- Mark BuildCacheServicesConfiguration.local/remote fields @Nullable (getters already
  @Nullable; a build may have no local/remote cache).
- OpFiringRemoteBuildCacheServiceHandle.getIOException: requireNonNull on the cause
  (the only constructor takes a non-null cause).
- requireNonNull on the AtomicReference results in DefaultBuildCacheController and
  BaseLocalBuildCacheServiceHandle, with TODOs referencing NullAway#681.

reinsch82 commented:
Rebased onto current `master` and retargeted the PR base to `master` (it was on `ml/nullaway-pt-next`, ~4874 commits behind). A literal replay wasn't viable — `master` had diverged and done overlapping nullability work — so I re-derived the enablement against current `master`, one commit per module. All 11 modules compile clean under NullAway and pass checkstyle. This also picks up `master`'s Gradle wrapper, so no separate Gradle upgrade is needed.

Your earlier review comments are incorporated:
- **ApiMemberSelector.classMember**: suppression + TODO for an `@Initializer`-style annotation.
- **SortingAnnotationVisitor.parentVisitor**: `@Nullable` + `requireNonNull` at the dereference (not a suppression).
- **TarBuildCacheEntryPacker**: `Objects.requireNonNull` instead of Guava `checkNotNull`.
- **OpFiringRemoteBuildCacheServiceHandle.getIOException**: `requireNonNull` on the (always-present) cause.
- **DefaultBuildCacheController / BaseLocalBuildCacheServiceHandle**: `requireNonNull` with TODOs referencing NullAway#681 (the `AtomicReference.get()` case).
- **TypeUtils**: `requireNonNull` instead of suppression.
- **NonHierarchicalFileWatcherUpdater**: plain `getMessage() != null` check, no suppression.
- **FileWatchingFilter**: suppression kept with a TODO referencing NullAway#681.
- **FieldMember**: no longer flagged on current master, so the suppression is gone entirely.

reinsch82 committed:
Keep WrapperConfiguration.getDistribution() nullable

A null distribution is a legitimate, documented state: it is what
WrapperExecutor.forProjectDirectory() yields when no wrapper properties
file exists, and DistributionFactory relies on it to fall back to the
current Gradle version. Returning requireNonNull(distribution) broke
WrapperExecutorTest and DistributionFactoryTest with NPEs.

Make the getter @Nullable (also on WrapperExecutor.getDistribution())
and move requireNonNull into the callers that genuinely require a
configured distribution: Install.createDist() and
PathAssembler.getDistribution().

Also update EXPECTED_WRAPPER_JAR_HASH in WrapperGenerationIntegrationTest:
the @Nullable annotations added to the wrapper classes on this branch
intentionally change the gradle-wrapper.jar bytecode.

reinsch82 committed:
Restore NullAway suppression in FieldMember.compareTo

The Linux CI toolchain no longer flags this call, but the Windows ARM
toolchain still rejects passing the @Nullable 'value' fields to
Comparator.compare() without an explicit @Nullable type argument, and
then rejects Ordering.arbitrary() (a Comparator<Object>) for the
@Nullable type variable. Bring back the explicit type argument plus the
justified suppression that was reviewed before the rebase.

reinsch82 committed:
Accept null in WrapperConfiguration.setDistribution()

The getter is @Nullable, and the accessor-symmetry rule in
KotlinCompatibilityTest (sanityCheck) requires the setter to match.
Does not affect the wrapper jar hash: R8 strips the annotation from
the minified gradle-wrapper.jar.

reinsch82 committed:
Remove NullAway #681 workarounds now that the issue is fixed

NullAway now infers that AtomicReference holds a non-null type (uber/NullAway#681),
so the requireNonNull() calls and @SuppressWarnings added for it are no longer needed:
- FileWatchingFilter.shouldWatchLocation: drop @SuppressWarnings.
- DefaultBuildCacheController / BaseLocalBuildCacheServiceHandle: return result.get()
  directly (drop requireNonNull and its now-unused import).

reinsch82 committed:
Remove remaining NullAway #681 workarounds

NullAway now infers that AtomicReference holds a non-null type (uber/NullAway#681),
so the pre-existing @SuppressWarnings("NullAway") added for it can go:
DefaultBuildOperationListenerManager, DefaultWorkerLeaseService, DefaultServiceRegistry
(x2) and InstrumentedInputs.listener() (keeping its @Deprecated).

reinsch82 committed:
Use an @Initializer annotation for ApiMemberSelector.classMember

Replace the @SuppressWarnings("NullAway") on the classMember field with a small
@Initializer annotation on the ASM visit() callback that populates it. NullAway
treats any annotation whose simple name is Initializer as marking a field
initializer, so the field is considered initialized without a blanket suppression.

mlopatkin commented on platforms/core-runtime/native/src/main/java/org/gradle/internal/nativeintegration/EnvironmentModificationResult.java:
```diff
@@ -16,20 +16,25 @@
     UNSUPPORTED_ENVIRONMENT("There is no native integration with this operating environment.");
 
+    @Nullable
     private final String reason;
```
:nail_care: Let's make it non-nullable, we can have `"Environment updated successfully"` as a value for `SUCCESS`.

mlopatkin commented on platforms/core-execution/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java:
```diff
@@ -521,7 +523,7 @@ private String getInternedFileName(Path path) {
 
         public FileSystemLocationSnapshot getResult() {
-            return builder.getResult();
+            return requireNonNull(builder.getResult());
```
:question: Why is it safe to assume that `getResult()` returns non-null here?

reinsch82 replied:
I changed it a bit, but in general the assumption was safe, since [this](https://github.com/gradle/gradle/blob/e8615a2d9a4a998cd23eade6980d915ca602e731/platforms/core-execution/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java#L196) also assumed `NonNull`. Now we through a dedicated exception in case it is null.

mlopatkin replied:
> I changed it a bit, but in general the assumption was safe, since [this](https://github.com/gradle/gradle/blob/e8615a2d9a4a998cd23eade6980d915ca602e731/platforms/core-execution/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java#L196) also assumed `NonNull`. Now we through an dedicated exception in case it is null.

IDEA shows [another callsite](https://github.com/gradle/gradle/blob/eef51ada5d868db32de254833a051ec69296d212/platforms/core-execution/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java#L403), which is `@Nullable` all the way up, so I was a bit reluctant about having a hard failure. Your current approach is fine by me, thank you for looking into it.

mlopatkin commented on platforms/core-execution/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java:
```diff
@@ -100,6 +100,8 @@ private List<FileSystemLocationFingerprint> fingerprintZipEntries(String zipFile
         }
     }
 
+    // The withInputStream callback below discards its result; the explicit null return is intentional.
```
:thinking: making `<T exteds @Nullable Object> withInputStream(...)` is probably a better solution here.

mlopatkin reviewed (changes requested):
Great to see another chunk of projects fixed! I have one concern about a suppression in DirectorySnapshotter, which I believe must be addressed before merging, and a couple of style nits.

mlopatkin commented on platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeClassSourceGenerator.java:
```diff
@@ -118,13 +122,14 @@ private static MethodSpec mapToMethodSpec(CallInterceptionRequest request, Consu
             return spec;
         } catch (Exception e) {
-            onFailure.accept(new HasFailures.FailureInfo(request, e.getMessage()));
+            onFailure.accept(new HasFailures.FailureInfo(request, requireNonNull(e.getMessage())));
```
⚠️ `getMessage()` is nullable for many exceptions (NPE and friends), and this is a generic `catch (Exception e)`. If a caught exception has no message, `requireNonNull` will throw here — from inside the failure handler — masking the original error with an unrelated NPE. Since we only need a message string, let's not require non-null: `e.toString()` (keeps the exception type) or `String.valueOf(e.getMessage())` would both be safe.

mlopatkin commented on platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java:
```diff
@@ -187,7 +191,7 @@ public Collection<Result> readRequest(ExecutableElement method, ReadRequestConte
                 .collect(Collectors.toList());
         } catch (IllegalArgumentException failure) {
-            return Collections.singletonList(new InvalidRequest(failure.getMessage()));
+            return Collections.singletonList(new InvalidRequest(requireNonNull(failure.getMessage())));
```
⚠️ Same concern as in `PropertyUpgradeClassSourceGenerator`: this only holds if every `IllegalArgumentException` reaching here was thrown by our own code with a non-null message. If one ever comes from a library, `requireNonNull` turns a reportable failure into an NPE. Safer to use `failure.toString()` / `String.valueOf(failure.getMessage())` rather than assert non-null.

reinsch82 committed:
Make EnvironmentModificationResult.reason non-nullable

Give SUCCESS a real message instead of null, as suggested in review.
This removes the @Nullable field and the @SuppressWarnings("NullAway")
on toString(), which could previously return null. The reason is only
printed in the !isSuccess() branch, so the new SUCCESS message never
appears in build output.

reinsch82 committed:
Allow nullable results from ZipEntry.withInputStream()

Bound the method's type parameter as <T extends @Nullable Object> in
ZipEntry and its implementations (FileZipInput, StreamZipInput,
FallbackHandlingResourceHasher), as suggested in review. The nested
IoFunction interface already declared @Nullable-bounded type parameters;
only the method's own type variable was non-null. This lets ZipHasher's
result-discarding callback return null without a NullAway suppression.

reinsch82 committed:
Use Exception.toString() for failure messages instead of requireNonNull

getMessage() is nullable for many exceptions (NPE and friends), and these
are generic catch blocks, so requireNonNull(getMessage()) could mask the
original failure with an unrelated NPE from inside the failure handler.
toString() is always non-null and keeps the exception type in the report.

reinsch82 committed:
Check DirectorySnapshotter's result at the API boundary

Make PathVisitor.getResult() honestly @Nullable instead of asserting
non-null with requireNonNull inside the visitor, and resolve the null
once in snapshot(), whose contract (the path points to an existing
directory) is what actually guarantees a result: the root is always
visited because PathVisitor.shouldVisitDirectory() never filters it.
If a future caller violates the contract, this now fails with a
descriptive IllegalStateException naming the path instead of an NPE
somewhere downstream in VFS code. The other getResult() call site,
followSymlink(), already returns @Nullable and needs no change.

reinsch82 committed:
Replace NullAway suppressions in NativeServices with null handling

Mark the static instance and nativeIntegration fields @Nullable and
handle the null cases explicitly instead of suppressing the check:
- Guard the service factory methods with nativeIntegration != null in
  addition to useNativeIntegrations.
- Inspect NativeException's cause and message null-safely in the
  constructor's fallback detection for already-loaded/locked
  native-platform libraries.

reinsch82 committed:
Make the @Initializer annotation public

reinsch82 committed:
Use the latest JDK 21 patch release for Windows ARM unit tests

The job was pinned to Microsoft OpenJDK 21.0.1 (October 2023). That javac
does not expose type-use annotations on symbols loaded from bytecode, so
NullAway's JSpecify mode cannot see the @Nullable bound on
ZipEntry.withInputStream()'s type parameter when compiling
:normalization-java against the :files jar, and rejects the null-returning
lambda in ZipHasher. Later 21.0.x patches carry the javac fix: the same
code compiles cleanly on Temurin 21.0.11 locally and on JDK 25 in all
other CI jobs.

Resolving '21' to the latest patch release fixes this and future
JSpecify-generics compilations on this job. The daemon pin to 21
(JDK-8371740 workaround) is unaffected.
